src/quicktype-core/language/CPlusPlus.ts

```diff
@@ -1082,62 +1257,56 @@ export class CPlusPlusRenderer extends ConvenienceRenderer {
         this.finishFile();
     }
 
-    protected emitGenerators(proposedFilename: string | undefined): void {
-        this.startFile("Generators.hpp", true);
-
-        if (proposedFilename !== undefined) {
-            this.emitLine(`#include "${proposedFilename}"`);
-            this.ensureBlankLine();
-        } else {
-            this.forEachNamedType(
-                "leading-and-interposing",
-                (c: ClassType, _n: Name) => this.emitIncludes(c),
-                (_e, _n) => {},
-                (u, _n) => this.emitIncludes(u)
-            );
-        }
-
+    protected emitMultiSourceStructure(proposedFilename: string): void {
         if (!this._options.justTypes && this.haveNamedTypes) {
-            this.emitNamespaces(["nlohmann"], () => {
-                this.forEachObject("leading-and-interposing", (c: ClassType, className: Name) =>
-                    this.emitClassFunctions(c, className)
-                );
+            this.emitHelper();
 
-                this.forEachEnum("leading-and-interposing", (e: EnumType, enumName: Name) =>
-                    this.emitEnumFunctions(e, enumName)
-                );
+            this.startFile("Generators.hpp", true);
 
-                if (this.haveUnions) {
-                    this.emitAllUnionFunctions();
-                }
+            this._allTypeNames.forEach(t => {
+                this.emitLine(`#include "${t}.hpp"`);
             });
-        }
 
-        this.finishFile();
-    }
+            this.ensureBlankLine();
+            this.emitGenerators();
 
-    protected emitMultiSourceStructure(proposedFilename: string): void {
-        this.emitHelper();
+            this.finishFile();
+        }
 
         this.forEachNamedType(
             "leading-and-interposing",
-            (c: ClassType, n: Name) => this.emitDefinition(c, n),
-            (e, n) => this.emitDefinition(e, n),
-            (u, n) => this.emitDefinition(u, n)
+            (c: ClassType, n: Name) => {
+                this.emitDefinition(c, n);  
+            },
+            (e, n) => {
+                this.emitDefinition(e, n);
+            },
+            (u, n) => {
+                this.emitDefinition(u, n);
+            }
         );
 
         /**
          * Quite a hack, this is ONLY to satisfy the test subsystem, which
          * explicitly looks for a Toplevel.hpp
          */
-        if (this._allTypeNames.size === 0) {
+        if (!this._generatedFiles.has(proposedFilename)) {
             this.startFile(proposedFilename);
+            this._generatedFiles.forEach(f => {
+                const include = (name: string): void => {
+                    this.emitLine(`#include "${name}"`);
+                };
+                include (f);
+            });
             this.finishFile();
         }
     }
 
     protected emitSourceStructure(proposedFilename: string): void {
+        this._generatedFiles.clear();
```

Comment:
Why is this necessary?  If `_generatedFiles` is used for two separate purposes, we should use two separate sets.